File: packages/expo-audio-stream/ios/Features.swift
//
//  Features.swift
//  ExpoAudioStream
//
//  Created by Arthur Breton on 23/6/2024.
//

import Foundation

public struct Features {
    var energy: Float
    var mfcc: [Float]
    var rms: Float
    var minAmplitude: Float
    var maxAmplitude: Float
    var zcr: Float
    var spectralCentroid: Float
    var spectralFlatness: Float
    var spectralRollOff: Float?
    var spectralBandwidth: Float?
    var chromagram: [Float]?
    var tempo: Float?
    var hnr: Float?
    var melSpectrogram: [Float]?
    var spectralContrast: [Float]?
    var tonnetz: [Float]?
    
    init(
        energy: Float = 0,
        mfcc: [Float] = [],
        rms: Float = 0,
        minAmplitude: Float = 0,
        maxAmplitude: Float = 0,
        zcr: Float = 0,
        spectralCentroid: Float = 0,
        spectralFlatness: Float = 0,
        spectralRollOff: Float? = nil,
        spectralBandwidth: Float? = nil,
        chromagram: [Float]? = nil,
        tempo: Float? = nil,
        hnr: Float? = nil,
        melSpectrogram: [Float]? = nil,
        spectralContrast: [Float]? = nil,
        tonnetz: [Float]? = nil
    ) {
        self.energy = energy
        self.mfcc = mfcc
        self.rms = rms
        self.minAmplitude = minAmplitude
        self.maxAmplitude = maxAmplitude
        self.zcr = zcr
        self.spectralCentroid = spectralCentroid
        self.spectralFlatness = spectralFlatness
        self.spectralRollOff = spectralRollOff
        self.spectralBandwidth = spectralBandwidth
        self.chromagram = chromagram
        self.tempo = tempo
        self.hnr = hnr
        self.melSpectrogram = melSpectrogram
        self.spectralContrast = spectralContrast
        self.tonnetz = tonnetz
    }
}

extension Features {
    func toDictionary() -> [String: Any] {
        return [
            "energy": energy,
            "mfcc": mfcc,
            "rms": rms,
            "minAmplitude": minAmplitude,
            "maxAmplitude": maxAmplitude,
            "zcr": zcr,
            "spectralCentroid": spectralCentroid,
            "spectralFlatness": spectralFlatness,
            "spectralRollOff": spectralRollOff ?? 0,
            "spectralBandwidth": spectralBandwidth ?? 0,
            "chromagram": chromagram ?? [],
            "tempo": tempo ?? 0,
            "hnr": hnr ?? 0,
            "melSpectrogram": melSpectrogram ?? [],
            "spectralContrast": spectralContrast ?? [],
            "tonnetz": tonnetz ?? []
        ]
    }
}

File: packages/expo-audio-stream/android/src/main/java/net/siteed/audiostream/Features.kt
package net.siteed.audiostream

import android.os.Bundle
import androidx.core.os.bundleOf

data class Features(
    val energy: Float = 0f,
    val mfcc: List<Float> = emptyList(),
    val rms: Float = 0f,
    val minAmplitude: Float = 0f,
    val maxAmplitude: Float = 0f,
    val zcr: Float = 0f,
    val spectralCentroid: Float = 0f,
    val spectralFlatness: Float = 0f,
    val spectralRollOff: Float = 0f,
    val spectralBandwidth: Float = 0f,
    val tempo: Float = 0f,
    val hnr: Float = 0f,
    val melSpectrogram: List<Float> = emptyList(),
    val chromagram: List<Float> = emptyList(),
    val spectralContrast: List<Float> = emptyList(),
    val tonnetz: List<Float> = emptyList(),
    val pitch: Float = 0f,
    val crc32: Long? = null
) {
    fun toDictionary(): Map<String, Any> {
        val baseMap = mapOf(
            "energy" to energy,
            "mfcc" to mfcc,
            "rms" to rms,
            "minAmplitude" to minAmplitude,
            "maxAmplitude" to maxAmplitude,
            "zcr" to zcr,
            "spectralCentroid" to spectralCentroid,
            "spectralFlatness" to spectralFlatness,
            "spectralRollOff" to spectralRollOff,
            "spectralBandwidth" to spectralBandwidth,
            "tempo" to tempo,
            "hnr" to hnr,
            "melSpectrogram" to melSpectrogram,
            "chromagram" to chromagram,
            "spectralContrast" to spectralContrast,
            "tonnetz" to tonnetz,
            "pitch" to pitch,
            "crc32" to (crc32 ?: 0)
        )
        return baseMap.filterValues { it != null }
    }

    fun toBundle(): Bundle {
        return bundleOf(
            "energy" to energy,
            "mfcc" to mfcc,
            "rms" to rms,
            "minAmplitude" to minAmplitude,
            "maxAmplitude" to maxAmplitude,
            "zcr" to zcr,
            "spectralCentroid" to spectralCentroid,
            "spectralFlatness" to spectralFlatness,
            "spectralRollOff" to spectralRollOff,
            "spectralBandwidth" to spectralBandwidth,
            "tempo" to tempo,
            "hnr" to hnr,
            "melSpectrogram" to melSpectrogram,
            "chromagram" to chromagram,
            "spectralContrast" to spectralContrast,
            "tonnetz" to tonnetz,
            "pitch" to pitch,
            "crc32" to (crc32 ?: 0)
        )
    }

    companion object {
        fun parseFeatureOptions(options: Map<*, *>?): Map<String, Boolean> {
            return options?.let { map ->
                mapOf(
                    "energy" to (map["energy"] as? Boolean ?: false),
                    "mfcc" to (map["mfcc"] as? Boolean ?: false),
                    "rms" to (map["rms"] as? Boolean ?: false),
                    "zcr" to (map["zcr"] as? Boolean ?: false),
                    "dB" to (map["dB"] as? Boolean ?: false),
                    "spectralCentroid" to (map["spectralCentroid"] as? Boolean ?: false),
                    "spectralFlatness" to (map["spectralFlatness"] as? Boolean ?: false),
                    "spectralRollOff" to (map["spectralRollOff"] as? Boolean ?: false),
                    "spectralBandwidth" to (map["spectralBandwidth"] as? Boolean ?: false),
                    "chromagram" to (map["chromagram"] as? Boolean ?: false),
                    "tempo" to (map["tempo"] as? Boolean ?: false),
                    "hnr" to (map["hnr"] as? Boolean ?: false),
                    "melSpectrogram" to (map["melSpectrogram"] as? Boolean ?: false),
                    "spectralContrast" to (map["spectralContrast"] as? Boolean ?: false),
                    "tonnetz" to (map["tonnetz"] as? Boolean ?: false),
                    "pitch" to (map["pitch"] as? Boolean ?: false),
                    "crc32" to (map["crc32"] as? Boolean ?: false)
                )
            } ?: emptyMap()
        }
    }
}
